### ISSUE TITLE
Thrift parser requires definition before a type is referenced

### ISSUE THREAD

**Issue by jlgale** [eleme/thriftpy#106]
Recent versions of the Apache thrift compiler allow out-of-order definitions within a file. For example:

```
struct A {
    1: optional B b;
}

struct B {
    1: optional i32 val;
}
```

However, the thriftpy parser, as of 0.2.0, does not allow this. In isolation this restriction isn't too annoying, but this breaks compatibility with thrift definitions which previously worked and it also is incompatible with thrift definitions that work with the Apache compiler.

**Comment by hit9** [comment]
Have a look, @lxyu , do we need this feature? To allow out-of-order type definitions.

**Comment by lxyu** [comment]
@hit9 I we don't need it internally, but if the upstream supports it, let's do it if it's not too hard.

**Comment by hit9** [comment]
This feature can enable definitions like this:

```
struct Node {
  1: required Node next_
}
```

**Comment by hit9** [comment]
The latest stable version of  apache thrift is v0.9.2, this feature is included in the development version v1.0-dev.

Besides, the C language dose not support this feature yet. (Amm.. I don't know why official thrift enables this..).

So.. this feature will not be considered until the official stable version released.
